Fix numerous key import and listing issues

ni4 opened the pull request:
This PR updates key processing code with ability to merge key data and to extract public key from the secret ones.
Also it updates key importing code to use these functions for more clear and correct key import.
So now:
- once secret key is imported then it's public part will be created and saved to public keyring
- if armored file contains multiple keys (including public/secret), then all of them will be imported
- during the key listing if there are no keys in public keyring then secret keyring will be displayed (which should not happen anymore, but, still could be usable).

Fixes #507
Fixes #650 
Fixes #722 
Fixes #740

randombit commented:
Looks like Ruby tests are crashing so I guess there is at least one bug ;)

ronaldtse commented:
Only the builds that include ruby-rnp fails. A typical failure log is provided here.

```
+pushd /home/travis/local-builds/ruby-rnp
~/local-builds/ruby-rnp ~/build/riboseinc/rnp
+env CI=false LD_LIBRARY_PATH=/home/travis/local-builds/botan-install/lib:/home/travis/local-builds/jsonc-install/lib:/home/travis/local-builds/rnp-install/lib bundle exec rspec
NOTE: Gem::Specification#has_rdoc= is deprecated with no replacement. It will be removed on or after 2018-12-01.
Gem::Specification#has_rdoc= called from /home/travis/local-builds/ruby-rnp/rnp.gemspec:24.
Rnp::Encrypt
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569293 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569294 seconds ago
  raises an error when adding a signer without a secret key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569294 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569295 seconds ago
  raises an error for a signer that is unable to sign
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569295 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569296 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569296 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569296 seconds ago
  requests the correct key when adding a signer
  with public-key + symmetric + signing
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569297 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569297 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569298 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569299 seconds ago
[signed_src_finish() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:937] signer's key not found
    can be decrypted with symmetric pass pass1
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569300 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569300 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569301 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569302 seconds ago
[signed_src_finish() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:937] signer's key not found
    can be decrypted with symmetric pass pass2
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569303 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569304 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569304 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569305 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569306 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569306 seconds ago
    raises an error when no password is provided
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569307 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569308 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569308 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569309 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569310 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569310 seconds ago
    can be decrypted by user2
#<UnboundMethod: Rnp#encrypt>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569311 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569312 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569312 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569313 seconds ago
[init_encrypted_src() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:1933] failed to obtain decrypting key or password
  raises an error when not provided the key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569314 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569314 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569314 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569315 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569316 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569316 seconds ago
[init_encrypted_src() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:1933] failed to obtain decrypting key or password
  requests the correct key for decrypting
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569317 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569317 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569317 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569318 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569319 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569319 seconds ago
[init_encrypted_src() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:1933] failed to obtain decrypting key or password
  raises an error when not provided a password
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569320 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569320 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569320 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569321 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569322 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569322 seconds ago
  decrypts to the original plaintext
#<UnboundMethod: Rnp#symmetric_encrypt>
  with a single password
    encrypts and decrypts
    raises an error with no password
[encrypted_decrypt_cfb_header() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:1192] checksum check failed
[init_encrypted_src() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:1933] failed to obtain decrypting key or password
    raises an error with a bad password
Rnp::Key
  1ED63EE56FADC34D
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569324 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569325 seconds ago
    has the correct fingerprint
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569325 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569326 seconds ago
    has the correct keyid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569326 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569326 seconds ago
    has the correct grip
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569327 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569327 seconds ago
    raises an error when requesting primary userid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569328 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569328 seconds ago
    has no userids
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569328 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569329 seconds ago
    is a primary key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569329 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569330 seconds ago
    is not a subkey
    #<UnboundMethod: Rnp::Key#each_userid>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569330 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569330 seconds ago
      returns an empty enumerator
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569331 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569331 seconds ago
      produces no userids
    #<UnboundMethod: Rnp::Key#public_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569332 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569332 seconds ago
      should equal true
    #<UnboundMethod: Rnp::Key#secret_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569333 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569333 seconds ago
      should equal false
    #<UnboundMethod: Rnp::Key#public_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569334 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569334 seconds ago
      should equal String
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569335 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569335 seconds ago
      should equal #<Encoding:ASCII-8BIT>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569335 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569336 seconds ago
      should eql "GPG"
    #<UnboundMethod: Rnp::Key#secret_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569336 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569337 seconds ago
      should raise Rnp::NoSuitableKeyError
    when the secret key has been loaded
      #<UnboundMethod: Rnp::Key#secret_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569337 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569338 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569338 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569338 seconds ago
        should equal true
      #<UnboundMethod: Rnp::Key#secret_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569339 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569339 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569340 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569340 seconds ago
        should equal String
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569340 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569341 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569341 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569342 seconds ago
        should equal #<Encoding:ASCII-8BIT>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569342 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569343 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569343 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569343 seconds ago
        should eql "GPG"
    #<UnboundMethod: Rnp::Key#json>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569344 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569344 seconds ago
      has the correct key type
Rnp::Key
  4BE147BB22DF1E60
    has the correct fingerprint
    has the correct keyid
    has the correct grip
    raises an error when requesting primary userid
    has correct userids
    is a primary key
    is not a subkey
    #<UnboundMethod: Rnp::Key#each_userid>
      correctly iterates userids
      correctly iterates userids w/block
    #<UnboundMethod: Rnp::Key#public_key_present?>
      should equal true
    #<UnboundMethod: Rnp::Key#secret_key_present?>
      should equal false
    #<UnboundMethod: Rnp::Key#public_key_data>
      should equal String
      should equal #<Encoding:ASCII-8BIT>
      should eql "GPG"
    #<UnboundMethod: Rnp::Key#secret_key_data>
      should raise Rnp::NoSuitableKeyError
    when the secret key has been loaded
      #<UnboundMethod: Rnp::Key#secret_key_present?>
        should equal true
      #<UnboundMethod: Rnp::Key#secret_key_data>
        should equal String
        should equal #<Encoding:ASCII-8BIT>
        should eql "G10"
    #<UnboundMethod: Rnp::Key#json>
      has the correct key type
Rnp::Key
  7BC6709B15C23A4A
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569345 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569345 seconds ago
    has the correct fingerprint
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569346 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569346 seconds ago
    has the correct keyid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569346 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569347 seconds ago
    has the correct grip
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569347 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569348 seconds ago
    has the correct primary userid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569348 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569349 seconds ago
    has correct userids
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569349 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569349 seconds ago
    is a primary key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569350 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569350 seconds ago
    is not a subkey
    #<UnboundMethod: Rnp::Key#each_userid>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569351 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569351 seconds ago
      correctly iterates userids
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569351 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569352 seconds ago
      correctly iterates userids w/block
    #<UnboundMethod: Rnp::Key#public_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569352 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569353 seconds ago
      should equal true
    #<UnboundMethod: Rnp::Key#secret_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569353 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569354 seconds ago
      should equal false
    #<UnboundMethod: Rnp::Key#public_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569354 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569354 seconds ago
      should equal String
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569355 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569355 seconds ago
      should equal #<Encoding:ASCII-8BIT>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569356 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569356 seconds ago
      should eql "GPG"
    #<UnboundMethod: Rnp::Key#secret_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569356 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569357 seconds ago
      should raise Rnp::NoSuitableKeyError
    when the secret key has been loaded
      #<UnboundMethod: Rnp::Key#secret_key_present?>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569357 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569358 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569358 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569359 seconds ago
        should equal true
      #<UnboundMethod: Rnp::Key#secret_key_data>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569359 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569359 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569360 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569360 seconds ago
        should equal String
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569361 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569361 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569362 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569362 seconds ago
        should equal #<Encoding:ASCII-8BIT>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569362 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569363 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569363 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569364 seconds ago
        should eql "GPG"
    #<UnboundMethod: Rnp::Key#json>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569364 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569364 seconds ago
      has the correct key type
Rnp::Key
  1ED63EE56FADC34D
    when loading a protected key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569365 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569365 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569366 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569366 seconds ago
      responds to inspect
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569367 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569367 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569367 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569368 seconds ago
      responds to to_s
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569368 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569369 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569369 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569369 seconds ago
      is protected
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569370 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569370 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569371 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569371 seconds ago
      is locked
      with no pass provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569372 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569372 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569372 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569373 seconds ago
        fails to unlock
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569373 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569374 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569374 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569374 seconds ago
        fails to unprotect
        when directly provided a bad pass
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569375 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569375 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569376 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569376 seconds ago
          fails to unlock
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569377 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569377 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569378 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569378 seconds ago
          fails to unprotect
      with a bad (string) pass provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569379 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569380 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569380 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569381 seconds ago
        fails to unlock
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569381 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569382 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569382 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569383 seconds ago
        fails to unprotect
      with a bad (proc) pass provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569384 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569384 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569385 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569385 seconds ago
        fails to unlock
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569386 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569387 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569387 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569387 seconds ago
        fails to unprotect
        when directly provided the correct pass
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569388 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569389 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569389 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569390 seconds ago
          unlocks
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569391 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569391 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569391 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569392 seconds ago
          remains protected after unlocking
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569393 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569393 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569394 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569394 seconds ago
          unprotects
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569395 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569395 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569396 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569396 seconds ago
          remains locked after unprotecting
    when loading an unprotected key
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569397 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569397 seconds ago
      is not protected
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569399 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569399 seconds ago
      is not locked
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569400 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569401 seconds ago
      can be locked
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569402 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569402 seconds ago
      can be protected
      with no pass provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569404 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569405 seconds ago
        can sign
#<Method: Rnp.default_homedir>
  returns the expected value
#<Method: Rnp.homedir_info>
  spec/data/keyrings/gpg
    returns the correct info
  spec/data/keyrings/gpg21
    returns the correct info
#<Method: Rnp.key_format>
  GPG
    should eql "GPG"
  KBX
    should eql "KBX"
  G10
    should eql "G10"
  Invalid
    should eql nil
Rnp
  has a version number
  responds to inspect
  #<UnboundMethod: Rnp#public_key_count>
    is 0 before loading any keys
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569406 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569407 seconds ago
    has the expected value after loading keys
  #<UnboundMethod: Rnp#secret_key_count>
    is 0 before loading any keys
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569407 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569408 seconds ago
    has the expected value after loading keys
  #<UnboundMethod: Rnp#find_key>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569408 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569409 seconds ago
    returns nil for keys that do not exist
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569409 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569410 seconds ago
[str_to_locator() /home/travis/build/riboseinc/rnp/src/lib/rnp2.cpp:2270] Invalid identifier type: badid
    raises an error for invalid search criteria
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569410 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569411 seconds ago
    finds keys by userid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569411 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569412 seconds ago
    finds keys by keyid
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569412 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569413 seconds ago
    finds keys by fingerprint
  #<UnboundMethod: Rnp#load_keys>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569413 seconds ago
    loads only public keys when specified
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569414 seconds ago
    loads only secret keys when specified
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569414 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569415 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569415 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569416 seconds ago
    loads all keys when specified
  #<UnboundMethod: Rnp#save_keys>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569416 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569416 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569417 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569417 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569418 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569418 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569419 seconds ago
    saves only public keys when specified
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569419 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569419 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569420 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569420 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569421 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569421 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569422 seconds ago
    saves only secret keys when specified
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569422 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569422 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569423 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569423 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569424 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569424 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569425 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569425 seconds ago
    saves all keys when specified (FAILED - 1)
  #<UnboundMethod: Rnp#userids>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569426 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569426 seconds ago
    has the expected count
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569427 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569427 seconds ago
    contains the expected values
    #<UnboundMethod: Rnp#each_userid>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569428 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569428 seconds ago
      returns an enumerator if not given a block
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569429 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569429 seconds ago
      correctly yields when given a block
  #<UnboundMethod: Rnp#keyids>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569430 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569430 seconds ago
    has the expected count
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569431 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569431 seconds ago
    contains the expected values
    #<UnboundMethod: Rnp#each_keyid>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569431 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569432 seconds ago
      returns an enumerator if not given a block
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569432 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569433 seconds ago
      correctly yields when given a block
  #<UnboundMethod: Rnp#fingerprints>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569433 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569434 seconds ago
    has the expected count
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569434 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569435 seconds ago
    contains the expected values
    #<UnboundMethod: Rnp#each_fingerprint>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569435 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569436 seconds ago
      returns an enumerator if not given a block
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569436 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569437 seconds ago
      correctly yields when given a block
  #<UnboundMethod: Rnp#grips>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569437 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569438 seconds ago
    has the expected count
    contains the expected values
  #<UnboundMethod: Rnp#generate_key>
    generating a lone primary
      only generated a single key
      has the correct type
      has the correct curve
      has the correct creation time
      has the correct usage
      has both public and secret parts
      has the correct userid
Rnp::Sign
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569441 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569442 seconds ago
  raises an error for keys that cannot sign
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569442 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569443 seconds ago
  raises an error for keys that have no secret key available
Rnp::Verify
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569443 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569443 seconds ago
  should equal false
#<UnboundMethod: Rnp#sign>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569445 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569445 seconds ago
  raises an error for keys that cannot sign
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569446 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569446 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails without a password provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569447 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569447 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails with an incorrect password
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569448 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569449 seconds ago
  returns the signature directly if provided no output
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569450 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569450 seconds ago
  uses the output specified, when provided
  #<UnboundMethod: Rnp#verify>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569451 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569451 seconds ago
    verifies a valid signature
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569453 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569453 seconds ago
[compressed_src_read() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:375] inflate error -4
[init_packet_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:2122] cannot read packet tag
    raises an error on a corrupted signature
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569455 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569455 seconds ago
[process_pgp_source() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:2251] not an OpenPGP data provided
    raises an error on invalid data
#<UnboundMethod: Rnp#cleartext_sign>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569457 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569457 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails without a password provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569458 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569458 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails with an incorrect password
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569459 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569459 seconds ago
  returns the signature directly if provided no output
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569460 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569461 seconds ago
  uses the output specified, when provided
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569462 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569462 seconds ago
  appears to generate a cleartext signature
  #<UnboundMethod: Rnp#verify>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569463 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569463 seconds ago
    verifies a valid signature
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569465 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569466 seconds ago
[armored_src_read() /home/travis/build/riboseinc/rnp/src/librepgp/stream-armor.cpp:238] wrong base64 character �
[armored_src_read() /home/travis/build/riboseinc/rnp/src/librepgp/stream-armor.cpp:238] wrong base64 character �
[signed_read_single_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:799] failed to read signature packet header
    raises an error on a corrupted signature
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569467 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569468 seconds ago
[process_pgp_source() /home/travis/build/riboseinc/rnp/src/librepgp/stream-parse.cpp:2251] not an OpenPGP data provided
    raises an error on invalid data
#<UnboundMethod: Rnp#detached_sign>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569469 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569469 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails without a password provider
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569470 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569470 seconds ago
[signed_fill_signature() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1088] wrong secret key password
[process_stream_sequence() /home/travis/build/riboseinc/rnp/src/librepgp/stream-write.cpp:1738] failed to finish stream
  fails with an incorrect password
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569471 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569472 seconds ago
  returns the signature directly if provided no output
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569473 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569473 seconds ago
  uses the output specified, when provided
  #<UnboundMethod: Rnp#detached_verify>
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569474 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569474 seconds ago
    verifies a valid signature
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569476 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569477 seconds ago
[signature_validate() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1043] wrong lbits
[signature_validate() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1043] wrong lbits
    raises an error on truncated data
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569478 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569479 seconds ago
[signature_validate() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1043] wrong lbits
[signature_validate() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1043] wrong lbits
    raises an error on modified data
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569480 seconds ago
[signature_check_binding() /home/travis/build/riboseinc/rnp/src/librepgp/stream-sig.cpp:1297] subkey expired 25569481 seconds ago
    raises an error on a corrupt signature
GC
./ci/main.sh: line 57:  6416 Aborted                 (core dumped) env CI=false LD_LIBRARY_PATH="${BOTAN_INSTALL}/lib:${JSONC_INSTALL}/lib:${RNP_INSTALL}/lib" bundle exec rspec
```

ni4 commented on src/librepgp/stream-key.cpp:
```diff
@@ -83,6 +82,54 @@ copy_signatures(list *dst, const list *src)
+}
+
+/**
+ * @brief Add signatures from src to list dst, skipping the dupliocates.
```
Thanks for noticing.

ni4 commented on src/librepgp/stream-packet.cpp:
```diff
@@ -1630,6 +1629,39 @@ copy_signature_packet(pgp_signature_t *dst, const pgp_signature_t *src)
+    if (memcmp(sig1->lbits, sig2->lbits, 2)) {
+        return false;
+    }
+
```
Agree, don't know I wrote it this way. Thanks.

ni4 commented:
@randombit  Thanks for the review. Yeah, ruby side stopped working. Will work on resolving this.

ni4 commented:
Ok, I found the source of problem. Actually it's not a bug but some kind of unfinished work (G10 keys went completely out of my scope). Will update this PR with some more commits later on.

ni4 commented:
Please see the updated PR. It will encounter problems with ruby-rnp till it is not updated.
@dewyatt What would be a good process to update together ruby-rnp checks & rnp? First commit at ruby-rnp side, or otherwise?

I changed the way keys are imported in FFI - now, once secret keys are loaded, their public parts are also added to the keyring.
@dewyatt @ronaldtse @randombit Do you agree on this?

dewyatt commented on src/lib/pgp-key.cpp:
```diff
@@ -217,6 +217,144 @@ pgp_key_free(pgp_key_t *key)
+        goto done;
+    }
+    // this data belongs to the key now
+    pktcp = (pgp_key_pkt_t){};
```
`pktcp = {}` will quiet GCC.

ni4 replied:
Thanks, fixed.

ni4 commented on src/tests/ffi.cpp:
```diff
@@ -471,8 +483,29 @@ test_ffi_save_keys(void **state)
+      rnp_load_keys(ffi, "GPG", input, RNP_LOAD_SAVE_PUBLIC_KEYS | RNP_LOAD_SAVE_SECRET_KEYS));
+    rnp_input_destroy(input);
+    input = NULL;
+    // check the counts. We should get only secret keys right now.
```
Thanks, fixed.
